### PR TITLE
fix: count() of `ATTACH` table should in sync with the original table

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

count() of `ATTACH` table should in sync with the original table

This is a bug in PR https://github.com/datafuselabs/databend/pull/13403 :

if a table is created by `attach table t ....  read_only` statement, `select count(*) from t` should reflects the changes that made to the table being attached. e.g.

~~~
mysql> create table base as select * from numbers(10);
Query OK, 10 rows affected (0.14 sec)

mysql> set global hide_options_in_show_create_table=0;
Query OK, 0 rows affected (0.02 sec)

mysql> show create table base;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                  |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| base  | CREATE TABLE `base` (
  `number` BIGINT UNSIGNED NOT NULL
) ENGINE=FUSE COMPRESSION='zstd' SNAPSHOT_LOCATION='2206/2211/_ss/1cb953bd2fd84614af2b7857929ccbe2_v4.mpk' STORAGE_FORMAT='parquet' |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.04 sec)
Read 0 rows, 0.00 B in 0.025 sec., 0 rows/sec., 0.00 B/sec.

mysql> attach table test_attach 's3://testbucket/admin/2206/2211/' connection=(access_key_id ='minioadmin' secret_access_key ='minioadmin' endpoint_url='http://127.0.0.1:9900/') read_only;
Query OK, 0 rows affected (0.06 sec)

mysql> select count() from test_attach;
+---------+
| count() |
+---------+
|      10 |
+---------+
1 row in set (0.05 sec)
Read 1 rows, 1.00 B in 0.004 sec., 225.45 rows/sec., 225.45 B/sec.

mysql> delete from base where number > 0;
Query OK, 9 rows affected (0.14 sec)

mysql> --  THIS IS NOT CORRECT, the `count()` should be 1
mysql> select count() from test_attach;
+--------+
| count() |
+--------+
|     10 |
+--------+
1 row in set (0.06 sec)
Read 1 rows, 8.00 B in 0.019 sec., 52.86 rows/sec., 422.87 B/sec.
~~~

This PR fixes the above issue.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13481)

### CODE REVIEW

File: src/query/storages/fuse/src/fuse_table.rs [hunk 1]
```diff
@@ -641,16 +641,38 @@ impl Table for FuseTable {
...
-            num_rows: Some(s.number_of_rows),
-            data_size: Some(s.data_bytes),
-            data_size_compressed: Some(s.compressed_data_bytes),
-            index_size: Some(s.index_data_bytes),
-            number_of_blocks: s.number_of_blocks,
-            number_of_segments: s.number_of_segments,
-        }))
+    async fn table_statistics(&self) -> Result<Option<TableStatistics>> {
+        let table_info = &self.table_info;
+        if Self::is_table_attached_read_only(&table_info.meta.options) {
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/fuse_table.rs:646]:
```suggestion
        match self.table_type {
          case ...
        }
```

Comment (dantengsky) [reply src/query/storages/fuse/src/fuse_table.rs]: addressed, hope I get it correctly.

File: src/query/storages/fuse/src/fuse_table.rs [hunk 2]
```diff
@@ -641,16 +641,40 @@ impl Table for FuseTable {
...
+                Ok(Some(TableStatistics {
+                    num_rows: Some(s.number_of_rows),
+                    data_size: Some(s.data_bytes),
+                    data_size_compressed: Some(s.compressed_data_bytes),
+                    index_size: Some(s.index_data_bytes),
+                    number_of_blocks: s.number_of_blocks,
+                    number_of_segments: s.number_of_segments,
+                }))
+            }
+        }
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/fuse_table.rs:677]:
```suggestion
 async fn table_statistics(&self) -> Result<Option<TableStatistics>> {
    match self.table_type {
        FuseTableType::AttachedReadOnly => {
            let snapshot = self.read_table_snapshot().await?
                .ok_or_else(|| ErrorCode::Internal(
                    "Failed to load snapshot of read_only attached table"
                ))?;

            let summary = &snapshot.summary;
            let statistics = TableStatistics {
                num_rows: Some(summary.row_count),
                data_size: Some(summary.uncompressed_byte_size),
                data_size_compressed: Some(summary.compressed_byte_size),
                index_size: Some(summary.index_size),
                number_of_blocks: Some(summary.block_count),
                number_of_segments: Some(snapshot.segments.len() as u64),
            };

            Ok(Some(statistics))
        }
        _ => {
            let s = &self.table_info.meta.statistics;
            let statistics = TableStatistics {
                num_rows: Some(s.number_of_rows),
                data_size: Some(s.data_bytes),
                data_size_compressed: Some(s.compressed_data_bytes),
                index_size: Some(s.index_data_bytes),
                number_of_blocks: s.number_of_blocks,
                number_of_segments: s.number_of_segments,
            };

            Ok(Some(statistics))
        }
    }
}

```

Comment (BohuTANG) [reply src/query/storages/fuse/src/fuse_table.rs:677]: GPT looks greate :/

Comment (dantengsky) [reply src/query/storages/fuse/src/fuse_table.rs:677]: indention issues? tried you suggestion, but seems cargo fmt does not agree ...

Comment (dantengsky) [reply src/query/storages/fuse/src/fuse_table.rs:677]: oh....  I see, thanks for your suggestion, that would be better

**Review by BohuTANG — APPROVED** [review]
LGTM

**Pull request merged**